Migrating Your GitHub CI to Hugging Face Jobs (#3403)
huggingface/blog: github-ci-hf-jobs.md

---
title: "Migrating Your GitHub CI to Hugging Face Jobs"
thumbnail: /blog/assets/github-ci-hf-jobs/thumbnail.gif
authors:
- user: abidlabs
---

# Migrating Your GitHub CI to Hugging Face Jobs

> 

Reviewer: We could add the thumbnail here as well (feel free to ignore)

If you have a GitHub repository and you have GitHub Actions enabled, you probably use GitHub-hosted runners for CI. That is the default for many projects because it is simple: add a workflow, write `runs-on: ubuntu-latest`, and GitHub gives you a machine.

That default is convenient, but it also has limits. GitHub Actions can be slow or down for maintenance, the hosted machines are generic, and GPU access is not something most open-source projects can just turn on. For [Trackio](https://github.com/gradio-app/trackio), those limits started to matter. We wanted both reliable CPU CI for basic unit tests and frontend checks, but also GPU CI for tests that need to run on actual CUDA hardware.

> So we tried an experiment: keep GitHub Actions as the CI control plane, but run selected jobs on [Hugging Face Jobs](https://huggingface.co/docs/hub/en/jobs-overview).

Reviewer: Should we lose the term "experiment" here? It feels like we are trying to push a matured product to the end user and we should word it like that. (i might be overthinking it, but as soon as I hear the term "experiment" I seem to have lower trust in the product thinking it is still in that stage)

The result: Trackio's CI now runs on Hugging Face Jobs and streams back real-time logs, **cutting our CI time for CPU jobs by about 30% and enabling a whole new test suite that runs on GPU machines**!

In this article, we explain step-by-step how to recreate the same setup for your GitHub repo. If you are using an agent, you can point it to this article, since we provide CLI instructions alongside browser-based instructions for us humans.

> Let's start with a quick intro to Hugging Face Jobs!

Reviewer: Maybe adding a Table of Contents here would be nice for the reader to quickly jump into the setup without having to scroll to the part?

## What is Hugging Face Jobs?

[Hugging Face Jobs](https://huggingface.co/docs/hub/en/jobs-overview) lets you run commands or scripts on Hugging Face's serverless infrastructure with almost any hardware flavor. A Job is essentially:

- a command to run
- a Docker image, from Docker Hub or a Hugging Face Space
- a hardware flavor, such as CPU or `t4-small` or `h200` GPU
- optional environment variables and secrets

> 

Reviewer 2: ```suggestion
A minimal example of running a job via the `hf` CLI looks like

```bash
hf jobs run <image> <command>
```
```

Reviewer 2: Maybe nice to have a quick example to give a sense of the syntax?

The key step is connecting GitHub Actions to HF Jobs, which we describe below.

## The architecture

For this setup, we created [`huggingface/jobs-actions`](https://github.com/huggingface/jobs-actions), a small bridge that turns a GitHub Actions job into an ephemeral self-hosted runner running inside an HF Job.

The complete flow looks like this:

1. A pull request triggers a GitHub Actions workflow.
1. GitHub queues any job whose `runs-on` label is not available, for example `hf-jobs-cpu-upgrade` or `hf-jobs-t4-small`, and sends a signed `workflow_job.queued` webhook to the dispatcher through the GitHub App.
1. The dispatcher Space verifies the webhook, checks for an `hf-jobs-*` label, mints a short-lived GitHub runner registration token, and starts an HF Job on the matching hardware.
1. The HF Job boots an ephemeral GitHub Actions runner and registers it with the repo using that one-shot token.

> 1. GitHub assigns the pending workflow job to that runner; the runner executes the CI job, reports status back to GitHub, and exits.

Reviewer: ```suggestion
1. A pull request triggers a GitHub Actions workflow.
2. GitHub queues any job whose `runs-on` label is not available, for example `hf-jobs-cpu-upgrade` or `hf-jobs-t4-small`, and sends a signed `workflow_job.queued` webhook to the dispatcher through the GitHub App.
3. The dispatcher Space verifies the webhook, checks for an `hf-jobs-*` label, mints a short-lived GitHub runner registration token, and starts an HF Job on the matching hardware.
4. The HF Job boots an ephemeral GitHub Actions runner and registers it with the repo using that one-shot token.
5. GitHub assigns the pending workflow job to that runner; the runner executes the CI job, reports status back to GitHub, and exits.
```

From GitHub's point of view, this is just a self-hosted runner. From Hugging Face's point of view, it is just a Job that launches a container to run the workflow steps from the repo’s GitHub Actions.

## Step 1: Duplicate the dispatcher Space

The first thing you need is the dispatcher. This is a small Docker Space that receives GitHub `workflow_job` webhook events and launches HF Jobs in response.

Create this first because the GitHub App needs a webhook URL, and that URL comes from the Space. This Space should be under your own namespace or under a Hugging Face org that you have write access to.

#### Web setup

Go to [`huggingface/jobs-actions-dispatcher`](https://huggingface.co/spaces/huggingface/jobs-actions-dispatcher) and click **Duplicate this Space**.

> <img width="996" alt="image" src="https://github.com/user-attachments/assets/c8b450c3-b801-43dc-97ff-954d9bbaf975" />

Reviewer: It is better to host the images of the blog post to https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog

```text
https://github.com/organizations/YOUR-GITHUB-ORG/settings/installations
```

## Step 3: Final dispatcher settings

At this point, the dispatcher Space should be configured. The GitHub App setup flow generated the commands that upload the App credentials, webhook secret, and Hugging Face token to the Space.

<img width="1317" alt="image" src="https://github.com/user-attachments/assets/0fc8ac73-f93a-419b-bd80-70da2756f50c" />

By default, HF Jobs are launched under the same namespace as the dispatcher Space. Optionally, set `HF_NAMESPACE` as a Space variable if you want to bill jobs to a different Hugging Face user or org:

```bash
export SPACE_ID=YOUR-HF-NAMESPACE/jobs-actions-dispatcher
hf spaces variables add "$SPACE_ID" -e HF_NAMESPACE=your-billing-namespace
hf spaces restart "$SPACE_ID"
```

The token you set in Step 2 should correspond to this namespace.

## Step 4: Change `runs-on`

The actual workflow change is small. Instead of:

```yaml
runs-on: ubuntu-latest
```

use one of the labels handled by the dispatcher:

```yaml
runs-on: hf-jobs-cpu-upgrade
```

For GPU tests, use a GPU label:

```yaml
runs-on: hf-jobs-t4-small
```

> For any GitHub Action you'd like to run on HF Jobs, this 1-line change is all you need!

Reviewer 2: This is very cool!

cat > .github/workflows/hf-jobs-test.yml <<'EOF'
name: HF Jobs Test

on:
  pull_request:
  push:
    branches: [main]
  workflow_dispatch:

jobs:
  test:
    runs-on: hf-jobs-cpu-upgrade
    steps:
      - uses: actions/checkout@v4
      - run: echo "Hello from Hugging Face Jobs"
EOF

git add .github/workflows/hf-jobs-test.yml
git commit -m "Run CI on Hugging Face Jobs"
git push
```

To verify from the CLI:

```bash
gh run list --repo YOUR-GITHUB-ORG/YOUR-REPO --limit 5
hf jobs ps --namespace "$HF_NAMESPACE"
hf spaces logs "$SPACE_ID"
```

You should be able to see logs just like a regular GitHub Action—for example, in this [Trackio PR #565](https://github.com/gradio-app/trackio/pull/565).

And that's it!

*Note on choosing the right Docker image*

Our first CPU setup used `ubuntu:22.04` and installed missing system packages during every run. That worked, but it was slower than it needed to be. GitHub's `ubuntu-latest` image includes a lot of developer tooling by default; a bare Ubuntu image does not.

For Trackio, the UI tests need Playwright browsers, Node, ffmpeg, sqlite, git, and normal Linux build dependencies. We tested the Microsoft Playwright image:

> 

Reviewer 2: Maybe link to https://huggingface.co/docs/hub/jobs-popular-images for more ML-related workloads, i.e. maybe people kick off training or batch inference via a GitHub action?

```

## Results

Here are the numbers from the Trackio PR experiment:

| Runner setup | Runtime | Compared to GitHub average |
| --- | ---: | ---: |
| GitHub `ubuntu-latest` baseline | `1m40s`  | baseline |
| HF Jobs CPU, Playwright image | `1m10s` | `-30s`, about `30%` faster |
| HF Jobs GPU, `t4-small` label | `45s` | no GitHub-hosted GPU baseline |

The biggest win was GPU CI. The Trackio GPU check ran on HF Jobs and passed in `45s`, costing less than a cent at the `t4-small` rate for that duration.

The CPU result was also encouraging. With the right image, the Linux test job was faster than the GitHub-hosted baseline. That suggests HF Jobs can be a practical CI backend, especially for ML projects that need custom images or accelerators.

Logs were another pleasant surprise. GitHub Actions logs are useful, but the web UI can be heavy for large logs. HF Jobs logs are easy to fetch from the CLI:

```bash
hf jobs logs <job_id> > logs.txt
```

That makes them easy to inspect with local tools or coding agents. In our bridge, we also mirrored the GitHub Actions job log into the HF Job log, so either system had enough information to debug a run.

> 

Reviewer 2: > [!TIP]
> If you just want to run Jobs on a schedule, you can also directly [schedule Jobs](https://huggingface.co/docs/hub/jobs-schedule).

---
title: "Migrating Your GitHub CI to Hugging Face Jobs"
thumbnail: /blog/assets/github-ci-hf-jobs/thumbnail.png
authors:
- user: abidlabs
---

# Migrating Your GitHub CI to Hugging Face Jobs

If you have a GitHub repository and you have GitHub Actions enabled, you probably use GitHub-hosted runners for CI. That is the default for many projects because it is simple: add a workflow, write `runs-on: ubuntu-latest`, and GitHub gives you a machine.

That default is convenient, but it also has limits. GitHub Actions can be slow, the hosted machines are generic, and GPU access is not something most open-source projects can just turn on. For [Trackio](https://github.com/gradio-app/trackio), those limits started to matter. We wanted normal CPU CI for unit tests and frontend checks, but we also wanted GPU CI so that tests which depend on actual CUDA hardware would fail when a GPU was missing or misconfigured.

So we tried an experiment: keep GitHub Actions as the CI control plane, but run selected jobs on [Hugging Face Jobs](https://huggingface.co/docs/hub/en/jobs-overview).

> The result: Trackio now has a CI setup where GitHub still owns the pull request checks, branch protection, and workflow UI, while Hugging Face Jobs provides the underlying CPU and GPU machines.

Author: ```suggestion
The result: Trackio's CI now runs on Hugging Face Jobs, cutting our CI time for CPU jobs by about 30% and enabling a whole new test suite that runs on GPU machines!
```
